### PR TITLE
resolves #143 - Joinmap and Bridge updated

### LINKED ISSUE #143: [FEATURE]-Update DisplayControllerJoinMap and DisplayBase to new Joinmap Scheme
**Is your feature request related to a problem? Please describe.**
New Version of essentials will be utilizing a new Joinmap schema - all of the devices need to eventually be converted, but this specific one needs to be converted "sooner" in order to facilitate documentation.

**Describe the solution you'd like**
Update the JoinMap and Bridge

**Describe alternatives you've considered**

**Additional context**
I'm updating essentials documentation - extant documentation references this joinmap specifically.  In order to maintain the topical relationship between this joinmap and the documentation, this needs to be udpated.

### PR DESCRIPTION
**Pull request by TrevorPayne**
Updated the JoinMap to the new AdvancedJoinmap and also linked up the bridge appropriately.

I'm concerned this may have some breaking changes for existing display plugins that may use the base device - since utilizing the "span" metadata for joins doesn't really jive with the old way of starting something at, say, 10, then incrementing by 1 for each new element.

It's been changed in this to start at 11, with a span of 10, for instance, to have joins from 11-20.  The end result is the same, the code is clearer with intent, but the implementation is slightly different.

If/When this is PR'd, I'll take it upon myself to update the existing display plugins to account for this change.

### CODE REVIEW

**Comment by andrew-welker** [comment]
This shouldn't break existing plugins that use the old style join maps that are in PepperDash.Essentials, not PepperDash.Essentials.Core. Those still exist, and will still work until those obsolete classes are removed at some point in the future. This really only affects internal devices, and will require whoever uses the internal devices to use the EiscBridgeAdvanced in their config instead of EiscBridge.

File: essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs [hunk 1]
```diff
@@ -179,32 +177,32 @@ public virtual FeedbackCollection<Feedback> Feedbacks
...
                 displayDevice.PowerOn();
             });
 
 
-            displayDevice.PowerIsOnFeedback.LinkInputSig(trilist.BooleanInput[joinMap.PowerOn]);
+            displayDevice.PowerIsOnFeedback.LinkInputSig(trilist.BooleanInput[joinMap.PowerOn.JoinNumber]);
 
-            var count = 1;
+            var count = 0;
             foreach (var input in displayDevice.InputPorts)
```
Comment (ndorin) [inline essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:194]: Here we likely need to perform some check that we don't exceed the span defined in the join map.  Might be better to use a for loop than foreach and first check what the displayDevice.InputPorts.Count value is and see if it's greater than the joinMap.InputSelectNumber.JoinSpan value

Comment (ndorin) [reply essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:194]: Also, the big clue here is that if we need a count variable outside the foreach, we should probably use a for loop instead.

Comment (TrevorPayne) [reply essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:194]: Im not sure there's an easy way to do a for loop here.  It's a dictionary.

I think the current implementation is fine as long as I add a check for the count as a function of the joinSpan.

`var count = 0;
foreach (var input in displayDevice.InputPorts)
{
    if (count < joinMap.InputSelectOffset.JoinSpan)
    {
        inputKeys.Add(input.Key);
        var tempKey = inputKeys.ElementAt(count - 1);
        trilist.SetSigTrueAction((ushort)(joinMap.InputSelectOffset.JoinNumber + count), () => displayDevice.ExecuteSwitch(displayDevice.InputPorts[tempKey].Selector));
        Debug.Console(2, displayDevice, "Setting Input Select Action on Digital Join {0} to Input: {1}", joinMap.InputSelectOffset.JoinNumber + count, displayDevice.InputPorts[tempKey].Key.ToString());
        trilist.StringInput[(ushort)(joinMap.InputNamesOffset.JoinNumber + count)].StringValue = input.Key.ToString();
        count++;
    }
}`

Comment (TrevorPayne) [reply essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:194]: Gross Formatting Just check my changes when I push it.

File: essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs [hunk 2]
```diff
@@ -1,281 +1,287 @@
...
+                {
+                    inputKeys.Add(displayDevice.InputPorts[i].Key);
+                    var tempKey = inputKeys.ElementAt(i);
+                    trilist.SetSigTrueAction((ushort)(joinMap.InputSelectOffset.JoinNumber + i),
+                        () => displayDevice.ExecuteSwitch(displayDevice.InputPorts[tempKey].Selector));
+                    Debug.Console(2, displayDevice, "Setting Input Select Action on Digital Join {0} to Input: {1}", 
+                        joinMap.InputSelectOffset.JoinNumber + i, displayDevice.InputPorts[tempKey].Key.ToString());
+                    trilist.StringInput[(ushort)(joinMap.InputNamesOffset.JoinNumber + i)].StringValue = displayDevice.InputPorts[i].Key.ToString();
+                }
+            }
```
Comment (ndorin) [inline essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:205]: This is looking good.  Can we add an else condition that prints a debug statement to warn users if they have a display with more inputs than the span?

Comment (TrevorPayne) [reply essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:205]: Will Do!

Comment (TrevorPayne) [reply essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:208]: Added a debug statement that says 
`"Device has {0} inputs.  The Join Map allows up to {1} inputs.  Discarding inputs {2} - {3} from bridge."` 
while displaying pertinent info, obviously

Comment (TrevorPayne) [reply essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:208]: I also want to note that I DID add MuteOn and MuteOff discrete controls to the joinMap and Bridge - they always get added when we build plugins anyway.  Shouldn't break anything.

Comment (ndorin) [reply essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:208]: This is SUPER close.  Only change I would make is to set the level for the debug statement to 0 and use the Debug.Console() overload that logs the message to the error log at the warning level as well.  This is the kind of thing that ins't just for debugging but we actually want to let the user take note of when they can't figure out why their input switching isn't working.

Comment (TrevorPayne) [reply essentials-framework/Essentials Core/PepperDashEssentialsBase/Display/DisplayBase.cs:208]: Done 

`Debug.Console(0, displayDevice, Debug.ErrorLogLevel.Warning, "Device has {0} inputs.  The Join Map allows up to {1} inputs.  Discarding inputs {2} - {3} from bridge.",
                        displayDevice.InputPorts.Count, joinMap.InputNamesOffset.JoinSpan, i + 1, displayDevice.InputPorts.Count);`

**Pull request merged**